src/main/resources/lib/nextjs/connection-config.es6

```diff
@@ -1,45 +1,95 @@
         applicationKey: app.name,
     });
 }
 
-export function getSiteConfig(pathOrId, repoId) {
+export function getSite(pathOrId, repoId) {
     const context = contextLib.get();
     if (context.repository !== repoId) {
         try {
             return contextLib.run({
                 principals: ["role:system.admin"],
                 repository: repoId,
             }, function () {
-                return getSiteConfigInContext(pathOrId);
+                return getSiteInContext(pathOrId);
             });
         } catch (e) {
             log.error('Failed to get site config: ' + e.message);
         }
     } else {
-        return getSiteConfigInContext(pathOrId);
+        return getSiteInContext(pathOrId);
     }
 }
 
-exports.getFrontendServerUrl = (config) => {
-    // read site config first
-    let url = config?.nextjsUrl;
+exports.getFrontendServerUrl = (site) => {
+    const projectName = getProjectName();
+    const siteName = site?._name;
+    let url;
+    if (app.name === APP_NEXTJS_NAME) {
+        // check the app-nextjs config for project and site
+        url = app?.config?.[`settings.${projectName}.${siteName}.url`];
+    }
     if (!url) {
-        // fall back to config file
-        url = app?.config?.nextjsUrl || "http://localhost:3000";
+        // fall back to third-party app using lib-nextjs config file
+        url = app?.config?.['nextjs.url'];
+    }
+    if (!url) {
+        // read site config next
+        url = getConfigFromSite(site)?.nextjsUrl || "http://localhost:3000";
     }
     return url.replace(removeEndSlashPattern, '');
 }
 
-exports.getFrontendServerToken = (config) => {
-    return config?.nextjsToken || app?.config?.nextjsToken;
+exports.getFrontendServerToken = (site) => {
+    const projectName = getProjectName();
+    const siteName = site?._name;
+
+    let token;
+    if (app.name === APP_NEXTJS_NAME) {
+        // check the app-nextjs config for project and site
+        token = app?.config?.[`settings.${projectName}.${siteName}.secret`];
+    }
+    if (!token) {
+        // fall back to third-party app using lib-nextjs config file
+        token = app?.config?.['nextjs.secret'];
+    }
+    if (!token) {
+        // read site config last
+        token = getConfigFromSite(site)?.nextjsToken;
+    }
+    return token;
+}
+
+const getProjectName = () => {
+    const context = contextLib.get();
+    let project = 'default';
+    if (context?.repository) {
+        project = context.repository.replace('com.enonic.cms.', '');
+    }
+    return project;
+}
+
+exports.getProjectName = getProjectName;
+
+const getConfigFromSite = (site) => {
+    const siteConfigs = site?.data?.siteConfig;
+    if (siteConfigs?.length) {
+        for (let i = 0; i < siteConfigs.length; i++) {
+            const datum = siteConfigs[i];
+            if (datum.applicationKey === app.name) {
+                return datum.config;
+            }
+        }
+    }
+    return undefined;
```

Comment:
You're on ES6, so you should use `Array.find`.

```
const getConfigFromSite = (site) => {
    const siteConfigs = site?.data?.siteConfig;
    if (!siteConfigs?.length) {
      return;
    }

    return siteConfigs.find(siteConfig => siteConfig.applicationKey === app.name)?.config;
}
```

Reply:
That doesn't work in nashorn unfortunately, but I will switch the if to bail out early